compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/HttpFiltersDesugar.java

```diff
@@ -151,25 +162,90 @@
+            return;
+        }
+        // Expected method type:
+        // `function authenticateResource(service object {} serviceRef)`
+        // The function is expected to panic with a distinct error when fail to authenticate.
+        // WebSocket listener will handle this error.
+        BSymbol methodSym = symResolver.lookupMethodInModule(packageSymbol, names.fromString(AUTHENTICATE_RESOURCE),
+                                                             env);
+        if (methodSym == symTable.notFoundSymbol || !(methodSym instanceof BInvokableSymbol)) {
+            return;
+        }
+        BInvokableSymbol invocationSymbol = (BInvokableSymbol) methodSym;
+        BLangResourceFunction resourceNode = (BLangResourceFunction) functionNode;
+
+        Location pos = resourceNode.getPosition();
+
+        // Create method invocation.
+        BLangSimpleVarRef selfRef = ASTBuilderUtil.createVariableRef(
+                pos, resourceNode.symbol.receiverSymbol);
+
+        ArrayList<BLangExpression> args = new ArrayList<>();
+        args.add(selfRef);
+
+        BLangInvocation invocationExpr =
+                ASTBuilderUtil.createInvocationExprForMethod(pos, invocationSymbol, args, symResolver);
         BLangSimpleVariableDef result = ASTBuilderUtil.createVariableDef(pos,
-                ASTBuilderUtil.createVariable(pos, "$temp$http$filter$result", symTable.anyType, invocationExpr, null));
+                ASTBuilderUtil.createVariable(pos, "$temp$websocket$desugar$result",
+                                              symTable.anyType, invocationExpr, null));
+        List<BLangStatement> statements = getFunctionBodyStatementList(functionNode);
         statements.add(0, result);
 
         BVarSymbol resultSymbol = new BVarSymbol(0, names.fromIdNode(result.var.name), env.enclPkg.packageID,
-                result.var.type,
-                resourceNode.symbol, pos, VIRTUAL);
+                                                 result.var.type, resourceNode.symbol, pos, VIRTUAL);
         resourceNode.symbol.scope.define(resultSymbol.name, resultSymbol);
         result.var.symbol = resultSymbol;
     }
 
-    private BPackageSymbol getHttpPackageSymbol(SymbolEnv env) {
+    private BPackageSymbol getPackageSymbol(SymbolEnv env, String packageName) {
+        // This resolves the package symbol when the code have an import relevant to the particular service
         for (BLangImportPackage pkg : env.enclPkg.imports) {
-            if (pkg.symbol.pkgID.orgName.value.equals(ORG_NAME) && pkg.symbol.pkgID.name.value.equals(PACKAGE_NAME)) {
+            if (pkg.symbol.pkgID.orgName.value.equals(ORG_NAME) && pkg.symbol.pkgID.name.value.equals(packageName)) {
                 return pkg.symbol;
             }
         }
+        // This resolves the package symbol when the code is at a submodule of the module which have the service
+        // definition. In that case there is no any import relevant to the particular service.
```

Comment:
This resolves the package symbol when the code is at a submodule of the module which have the listener definition.

Reply:
Updated with https://github.com/ballerina-platform/ballerina-lang/pull/31045/commits/f29c2b1d5fdd13cce00ba38fb2be681a07a5ea65